community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactions.java

```diff
@@ -115,50 +106,36 @@ public KernelTransactions( Locks locks,
     }
 
-    /**
-     * This is the factory that actually builds brand-new instances.
-     */
-    private final Factory<KernelTransactionImplementation> factory = new Factory<KernelTransactionImplementation>()
-    {
-        @Override
-        public KernelTransactionImplementation newInstance()
-        {
-            Locks.Client locksClient = locks.newClient();
-            LegacyIndexTransactionState legacyIndexTransactionState =
-                    new LegacyIndexTransactionStateImpl( indexConfigStore, legacyIndexProviderLookup );
-            KernelTransactionImplementation tx = new KernelTransactionImplementation(
-                    statementOperations, schemaWriteGuard,
-                    locksClient, hooks, constraintIndexCreator, transactionHeaderInformationFactory,
-                    transactionCommitProcess, transactionMonitor, legacyIndexTransactionState,
-                    localTxPool, Clock.SYSTEM_CLOCK, tracers.transactionTracer, storageEngine );
-
-            allTransactions.add( tx );
-
-            return tx;
-        }
-    };
-
     @Override
     public KernelTransaction newInstance()
     {
         assertDatabaseIsRunning();
-        return localTxPool.acquire().initialize( storageEngine.metaDataStore().getLastCommittedTransactionId() );
+
+        Locks.Client locksClient = locks.newClient();
+        LegacyIndexTransactionState legacyIndexTransactionState =
+                new LegacyIndexTransactionStateImpl( indexConfigStore, legacyIndexProviderLookup );
+
+        long lastTransactionIdWhenStarted = storageEngine.metaDataStore().getLastCommittedTransactionId();
+
+        KernelTransactionImplementation tx = new KernelTransactionImplementation( statementOperations, schemaWriteGuard,
+                locksClient, hooks, constraintIndexCreator, transactionHeaderInformationFactory,
+                transactionCommitProcess, transactionMonitor, legacyIndexTransactionState,
+                this, Clock.SYSTEM_CLOCK, tracers.transactionTracer, storageEngine, lastTransactionIdWhenStarted );
+
+        allTransactions.add( tx );
+
+        return tx;
     }
 
-    /**
-     * Global pool of transactions, wrapped by the thread-local marshland pool and so is not used directly.
-     */
-    private final LinkedQueuePool<KernelTransactionImplementation> globalTxPool
-            = new LinkedQueuePool<KernelTransactionImplementation>( 8, factory )
+    public void transactionClosed( KernelTransaction tx )
     {
-        @Override
-        protected void dispose( KernelTransactionImplementation tx )
+        boolean removed = allTransactions.remove( tx );
+        if ( !removed )
         {
-            allTransactions.remove( tx );
-            tx.dispose();
-            super.dispose( tx );
+            throw new IllegalStateException( "Transaction: " + tx + " is not present in the " +
+                                             "set of known active transactions: " + allTransactions );
         }
-    };
+    }
```

Comment:
Is it necessary to throw when the `remove` call returns `false`? I'm thinking a no-op is just as useful. We have an `assertTransactionOpen` call in the close method of `KernelTransaction`, which is where this method is called from. (Maybe it's called from other places as well, I don't know)